pkg/envtest/webhook.go

```diff
@@ -0,0 +1,358 @@
+	}
+
+	err = o.ModifyWebhookDefinitions(hookCA)
+	if err != nil {
+		return err
+	}
+
+	if err := createWebhooks(config, o.MutatingWebhooks, o.ValidatingWebhooks); err != nil {
+		return err
+	}
+
+	if err := WaitForWebhooks(config, o.ValidatingWebhooks, o.MutatingWebhooks, *o); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// WaitForWebhooks waits for the Webhooks to be available through API server
+func WaitForWebhooks(config *rest.Config,
+	validatingHooks []*admissionreg.ValidatingWebhookConfiguration,
+	mutatingWebhooks []*admissionreg.MutatingWebhookConfiguration,
+	options WebhookInstallOptions) error {
+
+	waitingFor := map[schema.GroupVersionKind]*sets.String{}
+
+	if len(validatingHooks) > 0 {
+		waitingFor[validatingHooks[0].GroupVersionKind()] = &sets.String{}
+	}
+	for _, hook := range validatingHooks {
+		waitingFor[hook.GroupVersionKind()].Insert(hook.Name)
+	}
+
+	if len(mutatingWebhooks) > 0 {
+		waitingFor[validatingHooks[0].GroupVersionKind()] = &sets.String{}
+	}
+	for _, hook := range mutatingWebhooks {
+		waitingFor[hook.GroupVersionKind()].Insert(hook.Name)
+	}
+
+	// Poll until all resources are found in discovery
+	p := &webhookPoller{config: config, waitingFor: waitingFor}
+	return wait.PollImmediate(options.PollInterval, options.MaxTime, p.poll)
+}
+
+// poller checks if all the resources have been found in discovery, and returns false if not
+type webhookPoller struct {
+	// config is used to get discovery
+	config *rest.Config
+
+	// waitingFor is the map of resources keyed by group version that have not yet been found in discovery
+	waitingFor map[schema.GroupVersionKind]*sets.String
+}
+
+// poll checks if all the resources have been found in discovery, and returns false if not
+func (p *webhookPoller) poll() (done bool, err error) {
+	// Create a new clientset to avoid any client caching of discovery
+	c, err := client.New(p.config, client.Options{})
+	if err != nil {
+		return false, err
+	}
+
+	allFound := true
+	for gvk, names := range p.waitingFor {
+		if names.Len() == 0 {
+			delete(p.waitingFor, gvk)
+			continue
+		}
+		for _, name := range names.List() {
+			var obj = &unstructured.Unstructured{}
+			obj.SetGroupVersionKind(gvk)
+			err := c.Get(context.TODO(), client.ObjectKey{
+				Namespace: "",
+				Name:      name,
+			}, obj)
+
+			if err == nil {
+				names.Delete(name)
+			}
+
+			if errors.IsNotFound(err) {
+				allFound = false
+			}
+			if err != nil {
+				return false, err
+			}
+		}
+	}
+	return allFound, nil
+}
+
+// setupCA creates CA for testing and writes them to disk
+func (o *WebhookInstallOptions) setupCA() ([]byte, error) {
+	hookCA, err := integration.NewTinyCA()
+	if err != nil {
+		return nil, fmt.Errorf("unable to set up webhook CA: %v", err)
+	}
+
+	hookCert, err := hookCA.NewServingCert()
+	if err != nil {
+		return nil, fmt.Errorf("unable to set up webhook serving certs: %v", err)
+	}
+
+	localServingCertsDir, err := ioutil.TempDir("", "envtest-serving-certs-")
```

Comment:
silly, but can we use os.Getwd() here instead of ""? It makes it immediately clear, and more specifically bubbles up permissions or other disk issues in testing environments. been bitten by this in the past. very unlikely flake, but it's clear and it hedges against us not understanding why that's happening (and getting a different error later when we try to write files)